### PR TITLE
Caching some properties for efficiency. See issue #657.

### PR DESCRIPTION
**Pull request by rhodrin**
Currently `_prepare_arguments` (located in `operator.py`) is fairly time consuming. Upon repeated calls to `op.apply()` many arguments are being 're-prepared' when there is no need. This begins to address this problem by caching certain properties e.t.c.

Note that this is a first step and more work is required to further reduce the overheads of repeated operator calls.

### CODE REVIEW

**Comment by FabioLuporini** [comment]
some flake8 complaints, but apart from that this is GTG. How much is the improvement?

File: devito/ir/support/space.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 from functools import reduce
 from operator import mul
 
+from cached_property import cached_property
+
```
Comment (FabioLuporini) [inline devito/ir/support/space.py]: nitpicking: no need for a blank line here.

The rule by which we organise imports is:
- first group are the imports from the Python library
- blank line
- second group are the important from third-party libraries
- blank line
- third group are the devito imports

Comment (rhodrin) [reply devito/ir/support/space.py]: Thanks for the explanation. Space removed.

Assuming the new 'basic' profiling is enabled, in a basic problem on a 21x21 grid with 10,000 calls to op.apply() these changes being the time spent in `_prepare_arguments` from ~70% down to 40%.

**Comment by FabioLuporini** [comment]
there are some PEP8 issues to be fixed here (see flake8 report on travis/jenkins), and then we're GTG

**Comment by rhodrin** [comment]
Hmm, yeah.
But this problem only seems to occur in yask based setups? I've been trying to reproduce a similar bug via a local test but am failing to do so...

**Comment by FabioLuporini** [comment]
needs rebasing on ``master`` before merging, but otherwise, assuming green CI, this looks GTG to me

**Comment by FabioLuporini** [comment]
> **if** I got it right (and it's a big if), one possible solution is to turn **all** `pop`s into `get`s in **all** `_arg_values` in devito, including those in `dimension.py` and those in `yask/`.
> To be even more rigid to avoid future messups, we could cast the return value of `_arg_defaults` to a `frozendict` (a dictionary that cannot be modified (by the callees)).

this thing I wrote time ago was BS, as we're indeed popping, but only from kwargs, not from the _arg_defaults' return value.

Merging...

**Pull request merged**